relayer/store/beefy.go

```diff
@@ -0,0 +1,141 @@
+package store
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+	merkletree "github.com/wealdtech/go-merkletree"
+	"golang.org/x/crypto/blake2b"
+	"golang.org/x/crypto/sha3"
+)
+
+type NewSignatureCommitmentMessage struct {
+	Payload                       [32]byte
+	ValidatorClaimsBitfield       []*big.Int
+	ValidatorSignatureCommitment  []byte
+	ValidatorPosition             *big.Int
+	ValidatorPublicKey            common.Address
+	ValidatorPublicKeyMerkleProof [][32]byte
+}
+
+type CompleteSignatureCommitmentMessage struct {
+	ID                               *big.Int
+	Payload                          [32]byte
+	RandomSignatureCommitments       [][]byte
+	RandomSignatureBitfieldPositions []*big.Int
+	RandomValidatorAddresses         []common.Address
+	RandomPublicKeyMerkleProofs      [][][32]byte
+}
+
+type Beefy struct {
+	ValidatorAddresses []common.Address
+	SignedCommitment   SignedCommitment
+}
+
+func NewBeefy(validatorAddresses []common.Address, signedCommitment SignedCommitment) Beefy {
+	return Beefy{
+		ValidatorAddresses: validatorAddresses,
+		SignedCommitment:   signedCommitment,
+	}
+}
+
+func (b *Beefy) BuildNewSignatureCommitmentMessage(valAddrIndex int) (NewSignatureCommitmentMessage, error) {
+	sig0ProofContents, err := b.GenerateMmrProofOffchain(valAddrIndex)
+	if err != nil {
+		return NewSignatureCommitmentMessage{}, err
+	}
+
+	// Update signature format (Polkadot uses recovery IDs 0 or 1, Eth uses 27 or 28, so we need to add 27)
+	sigVal := b.SignedCommitment.Signatures[valAddrIndex].Value
+	sig0Commitment := sigVal[:] // TODO: increment recovery ID properly
+
+	commitmentHash := blake2b.Sum256(b.SignedCommitment.Commitment.Bytes())
+
+	validatorClaimsBitfield := []*big.Int{big.NewInt(123)} // TODO: add bitfield stuff properly
+
+	msg := NewSignatureCommitmentMessage{
+		Payload:                       commitmentHash,
+		ValidatorClaimsBitfield:       validatorClaimsBitfield,
+		ValidatorSignatureCommitment:  sig0Commitment,
+		ValidatorPublicKey:            b.ValidatorAddresses[valAddrIndex],
+		ValidatorPosition:             big.NewInt(int64(valAddrIndex)),
+		ValidatorPublicKeyMerkleProof: sig0ProofContents,
+	}
+
+	return msg, nil
+}
+
+func (b *Beefy) GenerateMmrProofOffchain(valAddrIndex int) ([][32]byte, error) {
```

Comment:
GenerateMerkleProofOffchain not mmr

Reply:
Whoops, updated in https://github.com/Snowfork/polkadot-ethereum/pull/314/commits/9251d5b526bf056a4094d3f29f06e9dec2d0bfa6